lib/Runtime/Library/JavascriptArray.cpp

```diff
@@ -9152,44 +9152,55 @@ namespace Js
             uint32 fromIndex = static_cast<uint32>(fromVal);
             uint32 toIndex = static_cast<uint32>(toVal);
 
-            while (count > 0)
+            if (typedArrayBase)
             {
-                JS_REENTRANT(jsReentLock, BOOL hasItem = obj->HasItem(fromIndex));
-                if (hasItem)
+                // Typed array can't suddenly transform into one of the other types like JavascriptArray, so it gets its own faster loop.
+                while (count > 0)
                 {
-                    if (typedArrayBase)
-                    {
-                        Var val = typedArrayBase->DirectGetItem(fromIndex);
+                    Var val = typedArrayBase->DirectGetItem(fromIndex);
```

Comment:
please move HasItem and DeleteItem code here as well. Although you can omit that code when length == typedArrayBase->GetLength() (because in that case has and Delete does not make sense). But I leave that up to you.

Reply:
I moved HasItem/DeleteItem code up here and re-ran the microbenchmark you provided. With the HasItem step included, a separate loop was no faster than the original, which makes sense because even a very simple branch predictor could figure out that we were taking the same path every time through the loop. I like the idea of adding a faster path for the case where length is sensible though.